### PR TITLE
Tentatively convert Sortingwood to use CAPABILITY_ITEM_HANDLER.

### PR DESCRIPTION
**Pull request by noobanidus**
I've no idea if I've done it properly or not, but it seems to function
correctly with chests (& double chests, I believe that's compensated for
by only having the single handler shared between the two chests) and
Iron Chests.

As I've specifically fetched the handler for the "top" slot, it doesns't
take anything out of nearby furnaces, but there's every chance that the
code could mess around with the input slots of normal machines.

If it does, I'm not sure how to fix this. :-?

Can someone who actually understands this properly have a look at it? I didn't change any of the sorting code, just the way it was interacting with inventories (so it might be a little inefficient?). But... yeah.

### CODE REVIEW

File: src/main/java/twilightforest/block/BlockTFMagicLogSpecial.java [hunk 1]
```diff
@@ -186,40 +188,33 @@ private void doMinersTreeEffect(World world, BlockPos pos, Random rand) {
...
-			if (block instanceof BlockChest) {
-				chestInventory = ((BlockChest) block).getContainer(world, iterPos, true);
-			}
+			if (!te.hasCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, EnumFacing.UP)) continue;
 
-			TileEntity te = world.getTileEntity(iterPos);
-			if (te instanceof IInventory && !te.isInvalid()) {
-				teInventory = (IInventory) te;
-			}
+			handler = (IItemHandlerModifiable) te.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, EnumFacing.UP);
```
Comment (williewillus) [inline src/main/java/twilightforest/block/BlockTFMagicLogSpecial.java]: not every IItemHandler is IItemHandlerModifiable, so we can't assume IItemHandlerModifiable.

In general, you'll have to change the implementation from writing specific slots to just using the insert/extract helpers in `ItemHandlerHelper`

Comment (noobanidus) [reply src/main/java/twilightforest/block/BlockTFMagicLogSpecial.java]: Thanks! I didn't think it would be this simple, but I wasn't familiar enough with the code.

Is it safe to assume that all chest and chest equivalent and chest-adjacent chests would be IItemHandlerModifiable? Ergo, testing the interface to see if it is modifiable and skipping it if it isn't might be sufficient?

**Comment by noobanidus** [comment]
That feel when you don't have glasses on and don't notice the typo `git push -d` instead of `git push -f`

**Comment by noobanidus** [comment]
All right, so in addition to my commit message, here's a summary of my thoughts on the matter:

Most things that implement IItemHandlerModifiable are expecting that something will interact with their contents. The same cannot be said about IItemHandler. I feel like implementing the base IItemHandler would open up many more issues where the tree starts taking items from the input slots of machines, etc, in an over-zealous attempt to support.

Additionally, I defined a "minimum inventory size" of 10 (to be configurable hopefully). This avoids small inventories which happen to to implement IItemHandlerModifiable (such as furnaces) and can be accessed from the top from having their contents drained.

From my research, this means that:

### Originally/continually supported

* Vanilla chest & double chests
* Quark variants

### Newly supported

* Iron Chests (33m downloads)
* Actually Additions "barrel/chests" (18m downloads)
* Draconic Evolution's Draconic Chest (21m downloads)
* Stone Chest (80k downloads)
* MetalChests (60k downloads)
* Better Storage Too (50k downloads)
* Storage Cabinet (8k downloads)

Problematically would also be supported
* Refined Relocation (2m downloads)

Leading me to believe that, in addition to (once the other PR I have open for magical tree config is deemed suitable) adding configuration for the minimum inventory size, it would also make do to have a default tile entity blacklist for the sorting tree.

### Potentially supported

For all of the storage mods that I went through, the only one which used capabilities and did not use IItemHandlerModifiable was Colossal Chests, but the code was rather had to comprehend and spread across 3 separate Git repositories.

Additionally, Storage Drawers does use IItemHandler in its DrawerHandler, and I'm sure that custom casing for this might be desired given its large usage (30m downloads).

### Potentially "Losing" support

The following mods implement IInventory and, with the new switch to capabilities, would no longer be supported:

* Compact Storage (2m downloads, I haven't tested any of the mods to make sure it did actually work previously)
* Aroma1997's BetterChests (200k downloads)
* Stone Chest*s* (2k downloads)
* Spatial Systems (1.7k downloads)
* "Useful Chest" (500 downloads)

### Summary

I guess what I'm saying is: even without changing the code specifically to use IItemHandler (which I feel could become a problem rather than a helper as most storage-itemhandlers are IItemHandlerModifiable), and filtering specifically for IItemHandlerModifiable, we can bring Sortingwood support to potentially millions of players using popular mods. 

The loss would seem to be specifically Compact Storage as the only potential case where people might've been relying on the code.

That is assuming we don't take advantage of InvWrapper to wrap IInventory-specific code in the IItemHandlerModifiable, which is something that I have been looking at for the past few minutes but have yet to implement.

In summary I think:

* There's more to gain accepting *only* IItemHandlerModifiable than is lost by giving up on IInventory.
* There needs to be a tile entity blacklist to prevent cross-over/interference with things like Refined Relocation, and a configurable "minimum inventory size".
* There's the possibility of wrapping all other IInventory-only coded entities with InvWrapper, although there may be some other logistics required in that.
* Finally, I think that supporting IItemHandler directly may indirectly result in the Sortingwood Engine trying to interact with things that are, by their very nature, not specifically storage but just happen to have an inventory.

</waffle>

**Comment by williewillus** [comment]
I appreciate your carefulness and research, but I don't think it needs to be as complicated as you're making it. Here's how I'd go about it:

1. Only use IItemHandler, IItemHandlerModifiable is an implementation detail and I maintain that we shouldn't be relying on it at all. (unless there's a major difficulty reimplementing the algorithm with just IItemHandler/ItemHandlerHelper)
2. pass the NULL EnumFacing to getCapability to get a "sideless " inventory. Machines should return their internal buffer space for this, or whichever area makes the most sense as a "generally read-writeable scratch area"
3. If the inventory dosn't have a cap but is IInventory, wrap it using InvWrapper

**Comment by Tamaized** [comment]
@noobanidus Would you like to revisit this with a 1.18 target?

**Comment by noobanidus** [comment]
I believe this was already merged?
